metricbeat/helper/conversion.go

```diff
@@ -1,85 +1,168 @@
-/*
-The conversion functions take a key and a map[string]string. First it checks if the key exists and logs and error
-if this is not the case. Second the conversion to the type is done. In case of an error and error is logged and the
-default values is returned. This guarantees that also if a field is missing or is not defined, still the full metricset
-is returned.
-*/
 package helper
 
 import (
+	"fmt"
 	"strconv"
 
+	"github.com/elastic/beats/libbeat/common"
 	"github.com/elastic/beats/libbeat/logp"
 )
 
+// Schema describes how a map[string]string object can be parsed and converted into
+// an event. The conversions can be described using an (optionally nested) common.MapStr
+// that contains Conv objects.
+type Schema struct {
+	conversions common.MapStr
```

Comment:
instead of `common.MapStr` consider `map[string]Mapper`

where

```
type Mapper interface{
  Map( ... ) error
}

// implements Mapper on top of Conv
type Primitive struct {
  Conv
}

// implements Mapper interface for structure
type Object struct {
  map[string]Conv
}
```

This will add some type-safety + no need for type-switch in applySchemaToEvent, as (Mapper).Map basically implements applySchemaToEvent (Schema is basically `type Schema Object`.

This turns schema definition from `NewSchema(common.MapStr{ ... })` into `h.Schema{ ... }`.